File: task2.js
const dns = require('dns');
const portscanner = require('portscanner');
const readline = require('readline');
const { exec } = require('child_process');

// Function to resolve the IP address of a given hostname
function resolveIPAddress(hostname) {
    return new Promise((resolve, reject) => {
        dns.lookup(hostname, (err, address) => {
            if (err) {
                reject(err);
            } else {
                resolve(address);
            }
        });
    });
}

// Function to scan open ports for a given IP address
function scanOpenPorts(ip) {
    return new Promise((resolve, reject) => {
        portscanner.findAPortInUse(1, 65535, ip, (error, port) => {
            if (error) {
                reject(error);
            } else {
                resolve(port);
            }
        });
    });
}

// Function to detect the service running on a given port
function detectService(ip, port) {
    return new Promise((resolve, reject) => {
        // Using nmap for service detection
        exec(`nmap -p ${port} ${ip}`, (error, stdout, stderr) => {
            if (error) {
                reject(`Error: ${stderr}`);
            } else {
                resolve(stdout);
            }
        });
    });
}

// Create an interface for reading user input from the command line
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Ask the user to enter a website URL
rl.question('Please enter the website URL (e.g., example.com): ', (hostname) => {
    resolveIPAddress(hostname)
        .then(ip => {
            console.log(`IP address for ${hostname}: ${ip}`);
            return scanOpenPorts(ip).then(port => {
                console.log(`Open port: ${port}`);
                return detectService(ip, port);
            }).then(serviceInfo => {
                console.log(`Service Info: \n${serviceInfo}`);
            });
        })
        .catch(error => {
            console.error('Error:', error);
        })
        .finally(() => {
            rl.close();
        });
});

File: task2.py
import socket
import nmap

def get_ip_address(website):
    try:
        ip_address = socket.gethostbyname(website)
        return ip_address
    except socket.gaierror:
        print("Could not resolve the website.")
        return None

def scan_ports(ip_address):
    nm = nmap.PortScanner()
    nm.scan(ip_address, '1-1024')  # Scanning ports 1 to 1024, you can adjust the range as needed
    open_ports = nm[ip_address]['tcp'].keys()
    return open_ports, nm

def get_service_info(nm, ip_address, port):
    service_name = nm[ip_address]['tcp'][port]['name']
    service_version = nm[ip_address]['tcp'][port]['version']
    return service_name, service_version

def main():
    website = input("Enter the website (e.g., www.example.com): ")
    ip_address = get_ip_address(website)
    if ip_address:
        print(f"IP Address of {website}: {ip_address}")
        
        open_ports, nm = scan_ports(ip_address)
        if open_ports:
            print("Open Ports and Services:")
            for port in open_ports:
                service_name, service_version = get_service_info(nm, ip_address, port)
                print(f"Port: {port}, Service: {service_name}, Version: {service_version}")
        else:
            print("No open ports found.")
    else:
        print("Failed to retrieve IP address.")

if __name__ == "__main__":
    main()
